File: 백준/Gold/7576. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

  static int[][] box;
  static int days = 0;
  static int[] di = {0, 0, 1, -1};
  static int[] dj = {1, -1, 0, 0};
  static int m, n;
  static Queue<Tomato> queue = new LinkedList<>();

  public static void main(String[] args) throws IOException {

    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    m = Integer.parseInt(st.nextToken());
    n = Integer.parseInt(st.nextToken());

    box = new int[n][m];

    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      for (int j = 0; j < m; j++) {
        box[i][j] = Integer.parseInt(st.nextToken());
        if (box[i][j] == 1) {
          queue.offer(new Tomato(i, j, 0));
        }
      }
    }

    bfs();

  }

  private static void bfs() {

    int day = 0;

    while (!queue.isEmpty()) {
      Tomato cur = queue.poll();
      day = cur.day;

      for (int d = 0; d < 4; d++) {

        int nx = cur.i + di[d];
        int ny = cur.j + dj[d];

        if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
          if (box[nx][ny] == 0) {
            box[nx][ny] = 1;
            queue.offer(new Tomato(nx, ny, day + 1));
          }
        }
      }
    }

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (box[i][j] == 0) {
          System.out.println(-1);
          System.exit(0);
        }
      }
    }

    System.out.println(day);

  }



  static class Tomato {
    int i;
    int j;
    int day;

    Tomato(int i, int j, int day) {
      this.i = i;
      this.j = j;
      this.day = day;
    }
  }

}

File: 백준/Gold/7576. 토마토/토마토.kt
fun main() {
    val (m, n) = readln().split(" ").map { it.toInt() }
    val queue = ArrayDeque<Tomato>()
    val box = Array(n) { i ->
        val row = readln().split(" ").map { it.toInt() }.toIntArray()
        for (j in 0 until m) {
            if (row[j] == 1) {
                queue.add(Tomato(i, j, 0))
            }
        }
        row
    }

    fun bfs(): Int {
        val dx = intArrayOf(0, 0, 1, -1)
        val dy = intArrayOf(1, -1, 0, 0)

        var day = 0
        while (queue.isNotEmpty()) {
            val cur = queue.removeFirst()
            day = cur.day

            for (d in 0..3) {
                val nx = cur.x + dx[d]
                val ny = cur.y + dy[d]

                if (nx in 0 until n && ny in 0 until m) {
                    if (box[nx][ny] == 0) {
                        box[nx][ny] = box[cur.x][cur.y] + 1
                        queue.add(Tomato(nx, ny, cur.day + 1))
                    }
                }
            }
        }

        for (i in 0 until n) {
            for (j in 0 until m) {
                if (box[i][j] == 0) {
                    return -1
                }
            }
        }

        return day
    }

    print(bfs())
}

data class Tomato(val x: Int, val y: Int, val day: Int)